src/sources/aws_s3/sqs.rs

```diff
@@ -0,0 +1,517 @@
+                            key: s3_event.s3.object.key.clone(),
+                        })
+                    })
+                    .unwrap_or(Ok(()))
+            }
+            None => Ok(()),
+        }
+    }
+
+    async fn receive_messages(&self) -> Result<Vec<Message>, RusotoError<ReceiveMessageError>> {
+        self.sqs_client
+            .receive_message(ReceiveMessageRequest {
+                queue_url: self.queue_url.clone(),
+                max_number_of_messages: Some(10),
+                visibility_timeout: Some(self.visibility_timeout_secs),
+                ..Default::default()
+            })
+            .map_ok(|res| res.messages.unwrap_or_default())
+            .await
+    }
+
+    async fn delete_message(
+        &self,
+        receipt_handle: String,
+    ) -> Result<(), RusotoError<DeleteMessageError>> {
+        self.sqs_client
+            .delete_message(DeleteMessageRequest {
+                queue_url: self.queue_url.clone(),
+                receipt_handle,
+            })
+            .await
+    }
+}
+
+// https://docs.aws.amazon.com/AmazonS3/latest/dev/notification-content-structure.html
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "PascalCase")]
+struct S3Event {
+    records: Vec<S3EventRecord>,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct S3EventRecord {
+    event_version: String, // TODO compare >=
+    event_source: String,
+    aws_region: String,
+    event_name: S3EventName,
+
+    s3: S3Message,
+}
+
+#[derive(Clone, Debug)]
+struct S3EventName {
+    kind: String,
+    name: String,
+}
+
+// https://docs.aws.amazon.com/AmazonS3/latest/dev/NotificationHowTo.html#supported-notification-event-types
+//
+// we could enums here, but that seems overly brittle as deserialization would  break if they add
+// new event types or names
+impl<'de> Deserialize<'de> for S3EventName {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        use serde::de::Error;
+
+        let s = String::deserialize(deserializer)?;
+
+        let mut parts = s.splitn(2, ':');
+
+        let kind = parts
+            .next()
+            .ok_or_else(|| D::Error::custom("Missing event type"))?
+            .parse::<String>()
+            .map_err(D::Error::custom)?;
+
+        let name = parts
+            .next()
+            .ok_or_else(|| D::Error::custom("Missing event name"))?
+            .parse::<String>()
+            .map_err(D::Error::custom)?;
+
+        Ok(S3EventName { kind, name })
+    }
+}
+
+impl Serialize for S3EventName {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(&format!("{}:{}", self.name, self.kind))
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
```

Comment:
These three `rename_all` are unnecessary.

Reply:
They are, strictly, redundant, but I find them useful as a sort of documentation about the serialized form.

I'm just pulling out `bucket` and `object` here but there are additional keys, like `s3SchemaVersion`, that would need to be translated to camelCase.